scripts/gen_dockerfile_test.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/env python3
+
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit test for gen_dockerfile.py"""
+
+import argparse
+import os
+import re
+import shutil
+import subprocess
+import tempfile
+import unittest
+import unittest.mock
+
+import yaml
+
+import gen_dockerfile
+
+# Expected list of files generated
+EXPECTED_OUTPUT_FILES = ['Dockerfile', '.dockerignore']
+
+
+class GenDockerfileTest(unittest.TestCase):
+    def setUp(self):
+        self.testdata_dir = os.path.join(os.path.dirname(__file__), 'testdata')
+        assert os.path.isdir(self.testdata_dir), 'Could not run test: testdata directory not found'
+
+    def compare_file(self, filename, dir1, dir2):
+        """Compare identically named files in two different directories"""
+        with open(os.path.join(dir1, filename), 'r', encoding='utf8') as file1:
+            with open(os.path.join(dir2, filename), 'r', encoding='utf8') as file2:
+                contents1 = file1.read()
+                contents2 = file2.read()
+                msg = 'Contents of "{}" differ between "{}" and "{}"'.format(
+                    filename, dir1, dir2)
+                self.assertMultiLineEqual(contents1, contents2, msg)
+
+    def test_get_app_config(self):
+        args = argparse.Namespace(
+            config='some_config_file',
+            base_image='some_image_name',
+            source_dir='some_source_dir')
+
+        valid_cases = (
```

Comment:
Any time I can use regular code (`with unittest.subTest`) instead of a decorator (`@pytest.mark.parameterize`), I've done a good thing.  Decorators are code that runs invisibly, at global scope, at module import time.  Just an all-around misfeature of Python.